Sources/PenguinStructures/PCGRandomNumberGenerator.swift

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 Penguin Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Fast pseudorandom number generator using [permuted congruential
+/// generators](https://www.pcg-random.org/).
+///
+/// It combines two underlying 32-bit output PCG-XSH-RS random number generators
+/// to balance speed with `RandomNumberGenerator`'s requirement of 64-bit
+/// output.
+public struct PCGRandomNumberGenerator: RandomNumberGenerator {
+  var first: PCG_XSH_RS_32_Generator
+
+  var second: PCG_XSH_RS_32_Generator
+
+  public init(state: UInt64) {
+    self.init(seed: state, seq: 0xda3e_39cb_94b9_5bdb)
+  }
+
+  public init(seed: UInt64, seq: UInt64) {
+    self.init(seed1: seed, seed2: seed, seq1: seq, seq2: seq)
+  }
+
+  public init(seed1: UInt64, seed2: UInt64, seq1: UInt64, seq2: UInt64) {
+    let mask: UInt64 = ~0 >> 1;
+
+    var (stream1, stream2) = (seq1, seq2)
+
+    if stream1 & mask == stream2 & mask {
```

Comment:
Sorry, I'm not familiar with this construction. Although I understand what the code-as-written is doing, do you mind adding a comment explaining the larger context for this approach, what properties this provides, and what guarantees need to be upheld?